### PR TITLE
Use endpoint for cs3ref + use parent_id for relative paths in CS3Iface

### PR DESCRIPTION
**Pull request by dragotin**

### CODE REVIEW

File: src/core/cs3iface.py [hunk 1]
```diff
@@ -48,6 +48,17 @@ def getuseridfromcreds(token, _wopiuser):
     For the CS3 API case, this is just the token'''
     return token
 
+def getcs3reference(endpoint, filepath):
```
Comment (glpatcern) [inline src/core/cs3iface.py:51]: A cosmetic change as this is a "private" function, and the argument is a generic reference:
```suggestion
def _getcs3reference(endpoint, fileref):
```

File: src/core/cs3iface.py [hunk 2]
```diff
@@ -100,9 +104,11 @@ def statx(endpoint, fileid, userid, versioninv=0):
     return stat(endpoint, fileid, userid, versioninv)
 
 
-def setxattr(_endpoint, filepath, userid, key, value, lockid):
+def setxattr(endpoint, filepath, userid, key, value, lockid):
     '''Set the extended attribute <key> to <value> using the given userid as access token'''
-    reference = cs3spr.Reference(path=filepath)
+
+    # This assumes that the filepath is the same as the fileid like in the method above. Is that ok?
```
Comment (glpatcern) [inline src/core/cs3iface.py:110]: Now this assumption is generally false: the filepath is always a path in all methods other than `stat()`.

I understand this is a step towards supporting spaces as endpoints, but we have to see how the WOPI core logic must be adapted. The current logic uses `stat()` at the very beginning to resolve the file reference to a `(endpoint, absoulte_path)` in all cases, can we keep this assumption?

File: src/core/cs3iface.py [hunk 3]
```diff
@@ -63,15 +74,8 @@ def stat(endpoint, fileid, userid, versioninv=1):
...
-    if endpoint == 'default':
-        raise IOError('A CS3API-compatible storage endpoint must be identified by a storage UUID')
     tstart = time.time()
-    if fileid[0] == '/':
-        # assume this is a filepath
-        ref = cs3spr.Reference(path=fileid)
-    else:
-        # assume we have an opaque fileid
-        ref = cs3spr.Reference(resource_id=cs3spr.ResourceId(storage_id=endpoint, opaque_id=fileid))
+    ref = getcs3reference(endpoint, fileid)
```
Comment (glpatcern) [inline src/core/cs3iface.py:78]: Further to my other comment, we should rename `fileid` to `fileref` in the pre-existing code to be semantically accurate.

**Comment by glpatcern** [comment]
@dragotin and @wkloucek, shall we schedule some discussion to see how to proceed? I guess you're trying to support spaces, but the key point here is first to define how the `/openinapp` endpoint is (going to be) called from Reva within a space. Afterwards, we can see how to internally map the space id to an `endpoint`.

**Comment by glpatcern** [comment]
Update: following yesterday's "hackathon", this is feasible with some more extensions and after https://github.com/cs3org/cs3apis/pull/167 is implemented in Reva. Shouldn't take much.

File: src/core/cs3iface.py (hunk 2 above)
Comment (glpatcern) [reply src/core/cs3iface.py:109]: (in principle solved, see comment below)

**Comment by labkode** [comment]
@glpatcern the `parent_id` is now merged upstream and cs3apis binding should be there for Python. Note that we require Reva-side implementation as well to return this value.

**Comment by wkloucek** [comment]
@glpatcern please feel free to try https://github.com/cs3org/reva/pull/2691. That should unblock the WOPI server on REVA edge.

**Comment by glpatcern** [comment]
> @glpatcern please feel free to try [cs3org/reva#2691](https://github.com/cs3org/reva/pull/2691). That should unblock the WOPI server on REVA edge.

Thanks, looks good indeed but I won't have time to test it before Thursday I fear. Will do the equivalent code on Reva master for the eos provider - actually that PR could target Reva master and then get merged to Reva edge.

**Review by glpatcern — APPROVED** [review]

**Comment by wkloucek** [comment]
I tried it today with the latest oCIS and needed to merge master into this branch and apply following patch to make it work:

```diff
diff --git a/src/core/cs3iface.py b/src/core/cs3iface.py
index c51de8c..962ca42 100644
--- a/src/core/cs3iface.py
+++ b/src/core/cs3iface.py
@@ -95,8 +95,7 @@ def stat(endpoint, fileref, userid, versioninv=1):
         inode = common.encodeinode(statInfo.info.id.storage_id, statInfo.info.id.opaque_id)
         # in case we got a relative path, build an hybrid path that can be used to reference the file:
         # note that as per specs the parent_id MUST be available in this case
-        filepath = statInfo.info.path if statInfo.info.path[0] == '/' else \
-                        statInfo.info.parent_id.opaque_id + '/' + os.path.basename(statInfo.info.path)
+        filepath = statInfo.info.parent_id.opaque_id + '/' + os.path.basename(statInfo.info.path)
         log.info('msg="Invoked stat" fileref="%s" inode="%s" filepath="%s" elapsedTimems="%.1f"' %
                  (fileref, inode, filepath, (tend-tstart)*1000))
         return {
@@ -257,7 +256,7 @@ def readfile(endpoint, filepath, userid, lockid):
 
     # Download
     try:
-        protocol = [p for p in initfiledownloadres.protocols if p.protocol == "simple"][0]
+        protocol = [p for p in initfiledownloadres.protocols if p.protocol == "simple" or p.protocol == "spaces"][0]
         headers = {
             'x-access-token': userid,
             'x-reva-transfer': protocol.token        # needed if the downloads pass through the data gateway in reva
@@ -304,7 +303,7 @@ def writefile(endpoint, filepath, userid, content, lockid, islock=False):
     # Upload
     try:
         # Get the endpoint for simple protocol
-        protocol = [p for p in initfileuploadres.protocols if p.protocol == "simple"][0]
+        protocol = [p for p in initfileuploadres.protocols if p.protocol == "simple" or p.protocol == "spaces"][0]
         headers = {
             'x-access-token': userid,
             'Upload-Length': size,
```

Also I had to ensure that external locks are disabled (`detectexternallocks = False`). 

Then I could open files in edit mode, though I didn't succeed on saving files. This was most likely caused by the Set/GetXattr logic. From the logs I see that Set/GetXattr on REVA edge might be broken. Also I am not sure why a stat on current REVA edge returns a full path (it didn't the last time I checked)

**Comment by glpatcern** [comment]
@wkloucek thanks for this, you can definitely push the patch on `p.protocol == "spaces"`. Otherwise, the first patch won't be needed once Reva returns a relative path instead of a full path (yeah that does not sound correct), and the xattr logic must also work for save (PutFile) to work indeed.

**Review by glpatcern — APPROVED** [review]

**Pull request merged**